crates/lint/src/sol/mod.rs

```diff
@@ -0,0 +1,285 @@
+    pub(crate) fn lint_file(&self, file: &Path) -> Option<EmittedDiagnostics> {
+        let mut sess = if self.with_buffer_emitter {
+            Session::builder().with_buffer_emitter(ColorChoice::Never).build()
+        } else {
+            Session::builder().with_stderr_emitter().build()
+        };
+        sess.dcx = sess.dcx.set_flags(|flags| flags.track_diagnostics = false);
+
+        let arena = Arena::new();
+
+        let _ = sess.enter(|| -> Result<(), ErrorGuaranteed> {
+            // Declare all available passes and lints
+            let passes_and_lints: Vec<(Box<dyn EarlyLintPass<'_>>, SolLint)> = vec![
+                (Box::new(AsmKeccak256), ASM_KECCACK256),
+                (Box::new(IncorrectShift), INCORRECT_SHIFT),
+                (Box::new(DivideBeforeMultiply), DIVIDE_BEFORE_MULTIPLY),
+                (Box::new(VariableMixedCase), VARIABLE_MIXED_CASE),
+                (Box::new(ScreamingSnakeCase), SCREAMING_SNAKE_CASE),
+                (Box::new(StructPascalCase), STRUCT_PASCAL_CASE),
+                (Box::new(FunctionMixedCase), FUNCTION_MIXED_CASE),
+            ];
+
+            // Filter based on linter config
+            let mut passes: Vec<Box<dyn EarlyLintPass<'_>>> = passes_and_lints
+                .into_iter()
+                .filter_map(|(pass, lint)| {
+                    let matches_severity = match self.severity {
+                        Some(ref target) => target.contains(&lint.severity()),
+                        None => true,
+                    };
+                    let matches_lints_inc = match self.lints_included {
+                        Some(ref target) => target.contains(&lint),
+                        None => true,
+                    };
+                    let matches_lints_exc = match self.lints_excluded {
+                        Some(ref target) => target.contains(&lint),
+                        None => false,
+                    };
+
+                    if matches_severity && matches_lints_inc && !matches_lints_exc {
+                        Some(pass)
+                    } else {
+                        None
+                    }
+                })
+                .collect();
+
+            // Initialize the parser and get the AST
+            let mut parser = solar_parse::Parser::from_file(&sess, &arena, file)?;
+            let ast = parser.parse_file().map_err(|e| e.emit())?;
+
+            // Initialize and run the visitor
+            let ctx = LintContext::new(&sess, self.with_description);
+            let mut visitor = EarlyLintVisitor { ctx: &ctx, passes: &mut passes };
+            let _ = visitor.visit_source_unit(&ast);
+
+            Ok(())
+        });
+
+        sess.emitted_diagnostics()
+    }
+}
+
+impl Linter for SolidityLinter {
+    type Language = SolcLanguage;
+    type Lint = SolLint;
+
+    fn lint(&self, input: &[PathBuf]) {
+        input.into_par_iter().for_each(|file| {
```

Comment:
I think instead this we could pass the project and config to lint and then use `sess.enter_parallel` to filter and parse them instead `self.lint_file` for each file @DaniPopes wdyt?

Reply:
yes that would be better